### ISSUE TITLE
Transparent form with Skia

### ISSUE THREAD

**Issue by MichalMerco** [skia4delphi/skia4delphi#75]
Hi
First thing - FANTASTIC AND AWESOME WORK.
I wanted to create a transparent form (Transparency=True) in order to make a semi modal window on android (input box, message box etc). When I enable skia and set GlobalUseSkia in dpr transparency is lost. All you have to do is to enable Skia and put lines in dpr. After that form with transparency becomes black. Is there any workaround for that?

**Comment by viniciusfbb** [comment]
Hi @MichalMerco, thank you very much for reporting the issue.

There is no workaround at the moment, this issue is being reported for the first time.

To speed up our work, could you send us a project with the problem?

**Comment by MichalMerco** [comment]
Sure. I have attached a sample. Form with Transparency=True .
[skia.zip](https://github.com/skia4delphi/skia4delphi/files/8198308/skia.zip)

**Comment by MichalMerco** [comment]
Yes. I have downloaded today a new version. Look at the screenshots. Difference is only in line GlobalUseSkia 
![useSkiaFalse](https://user-images.githubusercontent.com/32061506/157054798-5be109e1-80e3-4834-bb46-cd9c1728904f.png)
![useSkiaTrue](https://user-images.githubusercontent.com/32061506/157054803-b580585a-92f0-4dce-a686-e495ea63525e.png)

**Comment by MichalMerco** [comment]
I am using Delphi 11

**Comment by viniciusfbb** [comment]
Are you running the project in a VM?

I couldn't reproduce the problem on my computer, I also tested it with custom scales and is working perfectly. Maybe this problem is related to issue #71

**Comment by MichalMerco** [comment]
Hi . On different machine with Delphi 10.4.2 everything works great. I have uninstalled FGX, UniGui and still no luck. So I suppose it is a problem with my Delphi 11. Anyway thanks for help and lets close this topic

**Comment by viniciusfbb** [comment]
Right. I'll close the issue but any news regarding the problem you can post here and I'll reopen.

**Comment by MichalMerco** [comment]
I guess problem is with my configuration. I have 2 monitors. When I disconnect secondary monitor (laptop only) transparency is fine. Problem is when secondary monitor is attached

**Comment by viniciusfbb** [comment]
> I guess problem is with my configuration. I have 2 monitors. When I disconnect secondary monitor (laptop only) transparency is fine. Problem is when secondary monitor is attached

But the problem occurs when you disable the Skia?

**Comment by MichalMerco** [comment]
No. When Skia is enabled and secondary monitor is attached through hdmi . And of course GlobalUseSkia=True. As soon as I detach monitor and start program on laptop everything is fine. Maybe problem is with GeForce

**Comment by MichalMerco** [comment]
Thx

**Comment by viniciusfbb** [comment]
Two monitors in extended mode working fine here:

![image](https://user-images.githubusercontent.com/11139086/157240893-c829563c-308e-4791-a234-6b27e0ac2a3d.jpeg)

**Comment by MichalMerco** [comment]
Thats right. In "secondary only" no

**Comment by viniciusfbb** [comment]
"Secondary only" mode works too.

**Comment by MichalMerco** [comment]
Well. Not for me. Look
![20220308_142354](https://user-images.githubusercontent.com/32061506/157246655-fd6158da-c984-4049-b7a7-b2660b601972.jpg)

**Comment by viniciusfbb** [comment]
One last important test is to get the project binaries (Project9.exe and sk4d.dll) and run it on another machine to check if the problem exists. Can you do this test or send me the binaries to test?

**Comment by MichalMerco** [comment]
I did that. On different machine even connected to secondary monitor (the same that my laptop is connected to) everything works. Problem is with my computer

**Comment by viniciusfbb** [comment]
Right. Now we have an idea of what it could be. We'll make some changes in the next release, and I'll let you know when we release the new version.

**Comment by MichalMerco** [comment]
Thank you very much

**Comment by MichalMerco** [comment]
Thank you very much 😊

 

Pozdrawiam

Michał Lutosławski

 

MERCO Systemy Komputerowe s.c. Leszek Pieńkowski, Michał Lutosławski

ul. Tetmajera 65C, 31-352 Kraków

tel:  792 77 77 54

NIP: 6772368217

**Comment by MichalMerco** [comment]
Thx

**Comment by birbilis** [comment]
> A possible workaround for both problems would be to use the CPU instead of the GPU, at least on Windows (on mobile this is still not possible). For that add the line GlobalUseSkiaRasterWhenAvailable := True;

if one doesn't set it (assuming the default is false for GlobalUseSkiaRasterWhenAvailable), does it mean it will try to use the GPU and fallback to CPU or will it just fail on older machines / graphics cards? (note that FMX can fallback even to GDI with GlobalUseDX := false). 

Also the "WhenAvailable" is a bit confusing, assuming it mean in platforms where the feature is available (since CPU is always there, it's GPU [drivers / DirectX or other SDK needed] that may be missing)

**Comment by viniciusfbb** [comment]
Hi @birbilis.

In the first versions the initial GlobalUseSkiaRasterWhenAvailable was false and from v3.3.2 the default became True. This was done because there are problems using our GPU canvas in some cases, especially on specific VMs that haven't added GPU support very well. There are few cases, but the problem is that we can't detect these cases at runtime to select the best option, so we leave CPU rendering at the default when available. We have implemented several skia-based canvas: Canvas Raster, Canvas OpenGL and Canvas Metal (in the future we will add Canvas Vulkan), but not all of them are available for each platform. The "when available" part means when there is a raster canvas implemented (since we only implement a raster canvas on windows and macOS)...

In the case of FMX, it uses Direct2D by default, which somehow manages to better switch CPU and GPU according to the senario. Unfortunately Skia only supports DirectX 12, which is present in Windows 10+, as we opted for OpenGL to maintain compatibility with Windows 7+.

Remembering that our raster mode (CPU) on Windows is as fast as the GPU mode, maybe 5% difference, something that in practice is not even possible to notice with the eye. The exception is shader operations, which run much faster on the GPU.

**Comment by birbilis** [comment]
Thanks for the informative answer
